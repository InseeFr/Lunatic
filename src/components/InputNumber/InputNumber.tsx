import { InputNumberThousand } from './InputNumberThousand';
import './InputNumber.scss';
import type { LunaticComponentProps } from '../type';
import { Label } from '../shared/Label/Label';
import {
	ComponentErrors,
	getComponentErrors,
} from '../shared/ComponentErrors/ComponentErrors';
import { slottableComponent } from '../shared/HOC/slottableComponent';
import { Declarations } from '../shared/Declarations/Declarations';
import type { LunaticError } from '../../use-lunatic/type';

export function InputNumber({
	handleChange,
	response,
	errors,
	decimals,
	...props
}: LunaticComponentProps<'InputNumber'>) {
	return (
		<CustomInputNumber
			{...props}
			decimals={decimals ?? 0}
			onChange={(v) => handleChange(response, v)}
			errors={getComponentErrors(errors, props.id)}
		/>
	);
}

type CustomProps = Omit<
	LunaticComponentProps<'InputNumber'>,
	'response' | 'handleChange' | 'errors'
> & {
	onChange: (v: number | null) => void;
	errors?: LunaticError[];
};

export const CustomInputNumber = slottableComponent<CustomProps>(
	'InputNumber',
	(props) => {
		const {
			id,
			value,
			onChange,
			disabled,
			readOnly,
			max,
			decimals,
			unit,
			label,
			errors,
			required,
			description,
			declarations,
		} = props;
		const labelId = `label-${id}`;
		return (
			<div className="lunatic-input-number">
				<Label htmlFor={id} id={labelId} description={description}>
					{label}
				</Label>
				<Declarations
					type="AFTER_QUESTION_TEXT"
					declarations={declarations}
					id={id}
				/>
				<InputNumberThousand
					id={id}
					value={value}
					onChange={onChange}
					disabled={!!disabled}
					readOnly={!!readOnly}
					required={!!required}
					labelId={labelId}
					max={max}
					decimals={decimals}
					invalid={!!errors}
				/>
				{unit && <span>{unit}</span>}
				<ComponentErrors errors={errors} />
			</div>
		);
	}
);
