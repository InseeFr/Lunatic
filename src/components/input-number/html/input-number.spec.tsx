import { fireEvent, render, screen } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import InputNumber from './input-number';
describe('InputNumber', () => {
	const mockOnChange = vi.fn();

	beforeEach(() => {
		mockOnChange.mockClear();
	});

	it('renders without crashing', () => {
		const container = render(
			<InputNumber
				value={'input' as any as number}
				id="input"
				aria-labelledby="input"
			/>
		);
		expect(container).toMatchSnapshot();
	});

	it('renders label and input', () => {
		const labelText = 'Enter a number';
		const { container } = render(
			<InputNumber id="number" label={labelText} onChange={mockOnChange} />
		);

		const label = screen.getByText(labelText);
		const input = container.querySelector('input[type="text"]');

		expect(label).toBeInTheDocument();
		expect(input).toBeInTheDocument();
	});

	it('renders with value', () => {
		const { container } = render(
			<InputNumber id="number" value={10} onChange={mockOnChange} />
		);

		const input = container.querySelector('input[type="text"]');

		expect(input).toHaveValue('10');
	});

	it('renders with big value', () => {
		const { container } = render(
			<InputNumber
				id="number"
				value={10000.45}
				decimals={2}
				onChange={mockOnChange}
			/>
		);

		const input = container.querySelector('input[type="text"]');

		expect(input).toHaveValue('10,000.45');
	});

	it('calls onChange with parsed value', () => {
		const { container } = render(
			<InputNumber id="number" onChange={mockOnChange} />
		);

		const input = container.querySelector('input[type="text"]')!;
		fireEvent.change(input, { target: { value: '10' } });

		expect(mockOnChange).toHaveBeenCalledTimes(1);
		expect(mockOnChange).toHaveBeenCalledWith(10);
	});

	it('renders unit element if provided', () => {
		const { getByText } = render(
			<InputNumber
				id="input-number-test"
				value={5}
				label="Input Number"
				labelId="input-number-label"
				unit="kg"
			/>
		);
		const unit = getByText('kg');
		expect(unit).toBeInTheDocument();
	});

	it('disables input when disabled prop is true', () => {
		const { container } = render(
			<InputNumber id="number" disabled onChange={mockOnChange} />
		);

		const input = container.querySelector('input[type="text"]');
		expect(input).toBeDisabled();
	});

	it('renders with unit', () => {
		const { container } = render(
			<InputNumber id="number" value={10} unit="kg" onChange={mockOnChange} />
		);

		const unit = container.querySelector('span');
		expect(unit).toHaveTextContent('kg');
	});
});
