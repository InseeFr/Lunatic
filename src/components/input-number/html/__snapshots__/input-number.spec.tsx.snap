// Vitest Snapshot v1

exports[`InputNumber > renders without crashing 1`] = `
<div>
  <div
    class="lunatic-input-number"
  >
    <input
      aria-invalid="false"
      aria-labelledby="label-input"
      class=""
      id="input"
      inputmode="numeric"
      lang="en"
      required=""
      step="1"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`InputNumber > should handle readOnly 1`] = `
<div>
  <div
    class="lunatic-input-number"
  >
    <input
      aria-invalid="false"
      aria-labelledby="label-number"
      class=""
      id="number"
      inputmode="numeric"
      lang="en"
      readonly=""
      required=""
      step="1"
      type="text"
      value="123"
    />
  </div>
</div>
`;
